classes/external/get_data_merge_requests.php

```diff
@@ -0,0 +1,142 @@
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+namespace tool_mergeusers\external;
+defined('MOODLE_INTERNAL') || die();
+global $CFG;
+require_once("{$CFG->libdir}/externallib.php");
+use \tool_mergeusers\merge_request;
+use external_function_parameters;
+use external_multiple_structure;
+use external_single_structure;
+use external_value;
+/**
+ * Version information
+ *
+ * @package     tool
+ * @subpackage  mergeusers
+ * @author      Nicola Vallinoto, Liguria Digitale
+ * @author      Jordi Pujol-Ahulló, SREd, Universitat Rovira i Virgili
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class get_data_merge_requests extends \external_api {
+    /**
+     * Webservice tool_mergeusers_get_data_merge_request parameters
+     *
+     * @return external_function_parameters
+     */
+    public static function execute_parameters(): external_function_parameters {
+        return new external_function_parameters([
+            'mergeusers' => new external_single_structure([
+                'removeuserfield' => new external_value(
+                    PARAM_TEXT, 'Remove user field', VALUE_OPTIONAL),
+                'removeuservalue' => new external_value(
+                    PARAM_RAW, 'Remove user value', VALUE_OPTIONAL),
+                'removeuserid' => new external_value(
+                    PARAM_INT, 'Identifier of the remove user id', VALUE_OPTIONAL),
+                'keepuserfield' => new external_value(
+                    PARAM_TEXT, 'Keep user field', VALUE_OPTIONAL),
+                'keepuservalue' => new external_value(
+                    PARAM_RAW, 'Keep user value', VALUE_OPTIONAL),
+                'keepuserid' => new external_value(
+                    PARAM_INT, 'Identifier of the keep user id', VALUE_OPTIONAL),
+                'id' => new external_value(
+                    PARAM_INT, 'Identifier of the merge request', VALUE_OPTIONAL),
+                'status' => new external_value(
+                    PARAM_INT, 'Status of the merge request', VALUE_OPTIONAL)
+            ])
+        ]);
+    }
+    /**
+     * Return data of the custom queue of the merge request
+     *
+     * @param $mergeusers
+     * @param $removeuserid
+     * @param $removeuservalue
+     * @param $keepuserid
+     * @param $keepuservalue
+     * @param $id
+     * @param $status
+     */
+    public static function execute(array $mergeusers) {
+        global $DB;
+        // Validate all of the parameters.
+        $params = array();
+        /*$params = self::validate_parameters(self::execute_parameters(),
```

Comment:
this have to be uncommented, as in the other web service.

Reply:
I did it and I received this error executing the web service from postman
    "exception": "invalid_parameter_exception",
    "errorcode": "invalidparameter",
    "debuginfo": "Missing required key in single structure: mergeusers"

Reply:
well, this is because you're not building properly the request. In your definition you have something like this:

```php
['mergeusers' => 
[
   'removeuserfield' => , // Optional.
   'removeuservalue' => , // Optional.
   ....
],
];
```

I showed it in PHP format, but it should be in JSON format, where all elements inside the element `mergeusers` is optional. For instance, in JSON format, it could be a request body like `{'mergeusers': {'removeuserfield': 'id', 'removeuservalue': '12345'}}`, so you are asking for every request queued for the given user id to remove.

If you want to simplify it and remove the need for the parent `mergeusers` item, you should define the [`parameters()`](https://github.com/jpahullo/moodle-tool_mergeusers/pull/242/files/4c881230aa7d1fa24a3b3a87cc49f5a190b717bf#diff-f33d929e72d007195accf1a37d4bc42a66280d9e6dc4ce339ee0903ad6d78d67R43-R58) method with all items as `VALUE_DEFAULT`, and adding a default value for every field, like an empty string or so. Internally, you should only consider for filtering on the SELECT those fields that are not empty string, for instance. This is for Moodle API, that defines that cannot be VALUE_OPTIONAL fields on the first level of parameters. 

As you wish.

Reply:
I'm having some troubles. After changing the definition as follows:

`public static function execute_parameters(): external_function_parameters {
        return new external_function_parameters([
            'mergeusers' => new external_single_structure([
                'removeuserfield' => new external_value(
                    PARAM_TEXT, 'Remove user field',  VALUE_DEFAULT, ''),
                'removeuservalue' => new external_value(
                    PARAM_RAW, 'Remove user value',  VALUE_DEFAULT, ''),
                'removeuserid' => new external_value(
                    PARAM_INT, 'Identifier of the remove user id',  VALUE_DEFAULT, 0),
                'keepuserfield' => new external_value(
                    PARAM_TEXT, 'Keep user field',  VALUE_DEFAULT, ''),
                'keepuservalue' => new external_value(
                    PARAM_RAW, 'Keep user value',  VALUE_DEFAULT, ''),
                'keepuserid' => new external_value(
                    PARAM_INT, 'Identifier of the keep user id',  VALUE_DEFAULT, 0),
                'id' => new external_value(
                    PARAM_INT, 'Identifier of the merge request',  VALUE_DEFAULT, 0),
                'status' => new external_value(
                    PARAM_INT, 'Status of the merge request',  VALUE_DEFAULT, 0)
            ])
        ]);
    }`

and calling the validate parameter as follows
`        $params = array();
        $params = self::validate_parameters(self::execute_parameters(), 
                            array('id' => $mergeusers->id));
`
I receive the following error

`{
    "exception": "invalid_parameter_exception",
    "errorcode": "invalidparameter",
    "message": "È stato rilevato un valore non valido di un parametro",
    "debuginfo": "Missing required key in single structure: mergeusers"
}`

Reply:
You are keeping the first level `mergeusers` item, as an array, you should have something like:

```php
return new external_function_parameters([
    'removeuserfield' => new external_value( PARAM_TEXT, 'Remove user field',  VALUE_DEFAULT, ''),
    'removeuservalue' => new external_value( PARAM_RAW, 'Remove user value',  VALUE_DEFAULT, ''),
    'removeuserid' => new external_value( PARAM_INT, 'Identifier of the remove user id',  VALUE_DEFAULT, 0),
    'keepuserfield' => new external_value( PARAM_TEXT, 'Keep user field',  VALUE_DEFAULT, ''),
    'keepuservalue' => new external_value( PARAM_RAW, 'Keep user value',  VALUE_DEFAULT, ''),
    'keepuserid' => new external_value( PARAM_INT, 'Identifier of the keep user id',  VALUE_DEFAULT, 0),
    'id' => new external_value( PARAM_INT, 'Identifier of the merge request',  VALUE_DEFAULT, 0),
    'status' => new external_value( PARAM_INT, 'Status of the merge request',  VALUE_DEFAULT, 0),
]); 
```

This should work now according to your test with postman.

Both ways are correct, but maybe the above one is simpler, since it removes the `mergeusers` useless key.

Reply:
No tested!

Reply:
changed as you suggested (removing the mergeusers useless key) but now I received a different message:
{
    "exception": "invalid_parameter_exception",
    "errorcode": "invalidparameter",
    "debuginfo": "Unexpected keys (mergeusers) detected in parameter array."
}

Reply:
@nvallinoto , you have to check how you use the your brand new Moodle web service now from postman. You are still using the key `mergeusers` there in the postman. Keep it trying and understand what you are invoking and how. You should not have any `mergeusers` key neither on your postman parameters. 

What I want to mean is that the old version, with `mergeusers` key and the current one, without it, are both valid. The problem exists on how you are invoking the service from your postman.

Reply:
the problem was in postman call where I invoked the array .. thanks